backend/arm64/isa/arm64_ast.ml

```diff
@@ -0,0 +1,597 @@
+  end
+
+  module Shift = struct
+    module Kind = struct
+      type t =
+        | LSL
+        | ASR
+        | LSR
+
+      let to_string t =
+        match t with LSL -> "lsl" | ASR -> "asr" | LSR -> "lsr"
+    end
+
+    type t =
+      { kind : Kind.t;
+        amount : Imm.t
+      }
+
+    let print ppf t =
+      let { kind; amount } = t in
+      Format.fprintf ppf "%s %a" (Kind.to_string kind) Imm.print amount
+  end
+
+  let print_separator ppf () = Format.fprintf ppf ", "
+
+  type label = string
+
+  module Addressing_mode = struct
+    (* CR gyorsh: only immediate offsets implemented. *)
+    type t =
+      | Offset of Reg.t * Imm.t
+      | Pre of Reg.t * Imm.t
+      | Post of Reg.t * Imm.t
+      | Literal of label
+
+    let print ppf t =
+      let open Format in
+      match t with
+      | Offset (r, imm) -> fprintf ppf "[%s, %a]" (Reg.name r) Imm.print imm
+      | Pre (r, imm) -> fprintf ppf "[%s, %a]!" (Reg.name r) Imm.print imm
+      | Post (r, imm) -> fprintf ppf "[%s], %a" (Reg.name r) Imm.print imm
+      | Literal l -> fprintf ppf "%s" l
+  end
+
+  type t =
+    | Imm of Imm.t
+    | Reg of Reg.t
+    | Extend of Extend.t
+    | Shift of Shift.t
+    | Sym of string
+    | Cond of Instruction_name.Cond.t
+    | Mem of Addressing_mode.t
+
+  let print ppf t =
+    let open Format in
+    match t with
+    | Imm imm -> Imm.print ppf imm
+    | Reg r -> Format.fprintf ppf "%s" (Reg.name r)
+    | Extend e -> Extend.print ppf e
+    | Shift s -> Shift.print ppf s
+    | Sym s -> fprintf ppf "%s" s
+    | Cond c -> Format.fprintf ppf "%s" (Instruction_name.Cond.to_string c)
+    | Mem m -> Format.fprintf ppf "%a" Addressing_mode.print m
+end
+
+module Instruction = struct
+  type t =
+    { name : Instruction_name.t;
+      operands : Operand.t array
+    }
+
+  let create name operands = { name; operands }
+
+  let print ppf t =
+    let { name; operands } = t in
+    let pp_sep = Operand.print_separator in
+    Format.fprintf ppf "%s\t%a"
+      (Instruction_name.to_string name)
+      (Format.pp_print_seq ~pp_sep Operand.print)
+      (Array.to_seq operands)
+end
+
+module Asm = struct
+  (** Emit assembly instructions for gas. *)
+
+  (* CR-soon gyorsh: The directives defined below duplicate [X86_ast]. Another
+     version of these directives is in [backend/asm_targets]. Refactor and
+     remove duplication. *)
+
+  type constant =
+    | Const of int64
+    | ConstThis
+    | ConstLabel of string
+    | ConstLabelOffset of string * int
+    | ConstAdd of constant * constant
+    | ConstSub of constant * constant
+
+  type line =
+    | Ins of Instruction.t
+    | Align of bool * int
+    | Byte of constant
+    | Bytes of string
+    | Comment of string
+    | Global of string
+    | Protected of string
+    | Hidden of string
+    | Weak of string
+    | Long of constant
+    | NewLabel of string
+    | NewLine
+    | Quad of constant
+    | Section of string list * string option * string list * bool
```

Comment:
(I understand the other version may
be defined like that, but an inline
record would be nicer.)

Reply:
I won't address this and the next comment to make it easier to identify the duplication during the upcoming refactoring of this area.

Reply:
Sure, but please add an item specifically
for this case - the generic one is not very
likely to be addressed soon.

Reply:
Added a CR too.